pyartcd/pyartcd/pipelines/ocp4_scan.py

```diff
@@ -0,0 +1,107 @@
+import asyncio
+import os
+import yaml
+
+import click
+
+from pyartcd import exectools, util
+from pyartcd.cli import cli, click_coroutine, pass_runtime
+from pyartcd.runtime import Runtime
+from pyartcd.jenkins import trigger_ocp4, trigger_rhcos, trigger_build_sync
+
+DOOZER_WORKING = 'doozer_working'
+
+
+class Ocp4ScanPipeline:
+
+    def __init__(self, runtime: Runtime, version: str):
+        self.runtime = runtime
+        self.version = version
+        self.logger = runtime.logger
+        self.rhcos_changed = False
+        self.rhcos_inconsistent = False
+        self.changes = {}
+
+    async def run(self):
+        self.logger.info('Building: %s', self.version)
+
+        # Check for RHCOS changes and inconsistencies
+        await asyncio.gather(*[self._get_changes(), self._rhcos_inconsistent()])
+
+        # Handle source changes, if any
+        if self.changes.get('rpms', None) or self.changes.get('images', None):
+            self.logger.info('Detected source changes:\n%s', yaml.safe_dump(self.changes))
+
+            if self.runtime.dry_run:
+                self.logger.info('Would have triggered a %s ocp4 build', self.version)
+                return
+
+            # Trigger ocp4
+            self.logger.info('Triggering a %s ocp4 build', self.version)
+            await trigger_ocp4(self.version)
+
+        elif self.rhcos_inconsistent:
+            if self.runtime.dry_run:
+                self.logger.info('Would have triggered a %s RHCOS build', self.version)
+                return
+
+            # Inconsistency probably means partial failure and we would like to retry.
+            #  but don't kick off more if already in progress.
+            self.logger.info('Triggering a %s RHCOS build for consistency', self.version)
+            await trigger_rhcos(self.version, True)
+
+        elif self.rhcos_changed:
+            if self.runtime.dry_run:
+                self.logger.info('Would have triggered a %s build-sync build', self.version)
+                return
+
+            self.logger.info('Triggering a %s build-sync', self.version)
+            await trigger_build_sync(self.version)
+
+    async def _get_changes(self):
+        """
+        Check for changes by calling doozer config:scan-sources
+        Changed rpms, images or rhcos are recorded in self.changes
+        self.rhcos_changed is also updated accordingly
+        """
+
+        # Run doozer scan-sources
+        cmd = f'doozer --assembly stream --working-dir={DOOZER_WORKING} --group=openshift-{self.version} ' \
+              f'config:scan-sources --yaml --ci-kubeconfig {os.environ["KUBECONFIG"]}'
```

Comment:
If `KUBECONFIG` is required, would like to add a check at the beginning of this job to fail earlier.

Reply:
Added an explicit check in 6542f9863c7a76e393cf47250ec816828e90315b
The code will raise a RuntimeError if `KUBECONFIG` is not defined